feature(module): Use browserstack-runner as module as well as cli

vedharish opened the pull request:
We had a support request to allow browserstack-runner to be used as a module so as to be able to `require('browserstack-runner')`. 

### Dependencies
- [x] Test: Added basic scenarios for testing invalid config keys and test report.

Sample code to use this as module -

```node
var browserstackRunner = require('browserstack-runner');

var config = require('./browserstack.json');

browserstackRunner.run(config, function(error, report) {
  if(error) {
    console.log("Error:" + error);
    return;
  }
  console.log(JSON.stringify(report, null, 2));
  console.log("Test Finished");
});
```

vedharish commented:
@jzaefferer @tr4n2uil Can you please have a look at the code? And how should we proceed here so as to ensure nothing breaks?

Should I start adding tests for all possible error scenarios and hope to cover enough cases? Also any cases where this will definitely break?

tr4n2uil commented:
Looks awesome to me. Am plus one for this.

jzaefferer commented:
The API is weird. Passing a filename for the json config doesn't make sense, I'd expect the config to be passed instead. That allows a lot more flexibilty and the current behaviour is easy enough to emulate:
```
var config = require('./browserstack.json')
var browserstackRunner = require('browserstack-runner');
browserstackRunner.run(function(error) {
  if(error) {
    console.log("Error:" + error);
    return;
  }
  console.log("Test Finished");
});
```
I also renamed `test` to `run` and added a return statement in the error handler.

Whatever the end result, that usage example should be added to the README as well.

As for adding tests, I can't help with that.

vedharish commented:
@jzaefferer @tr4n2uil 

Updated the usage as follows -
```node
var browserstackRunner = require('browserstack-runner');

global.logLevel = 'info';
var config = require('./browserstack.json');

browserstackRunner.run(config, function(error, report) {
  if(error) {
    console.log("Error:" + error);
    return;
  }
  console.log(JSON.stringify(report, null, 2));
  console.log("Test Finished");
});
```

Also added the `report` object in `run` callback and added its structure and the usage as a module in the readme [here](https://github.com/browserstack/browserstack-runner/blob/43c2b6e55117643cdf5ce9df6d103eadbf888ee9/README.md#usage-as-a-module)

Also added a few tests to check `reporting` and `error handling`

jzaefferer commented on README.md:
````diff
@@ -21,6 +21,99 @@ If you're getting an error `EACCES open ... BrowserStackLocal`, configure npm to
+```node
+var browserstackRunner = require('browserstack-runner');
+
+global.logLevel = 'info';
````
This is kinda strange. Should at least mention what this does and what other log levels are available.

jzaefferer commented on README.md:
```diff
@@ -21,6 +21,99 @@ If you're getting an error `EACCES open ... BrowserStackLocal`, configure npm to
+1. `error`: This parameter is either `null` or an `Error` object (if test execution failed) with message as the reason of why executing the tests on `BrowserStack` failed.
+2. `report`: This is an object which can be used to keep track of the `failed assertions` and the total count of `passed/failed` tests specific to a browser instance.
+
+The structure of the `report` object is as follows -
```
Instead of rolling your own, would you consider using the js-reporters format for this? https://github.com/js-reporters/js-reporters#event-data

I'd use an array for the top level structure, with the "browser" name as one property and the details as another.

akhillb commented on lib/config.js:
```diff
@@ -1,100 +1,84 @@
-    //Backward Compatibility, if test_path is not array of path
-    config.test_path = formatPath(config.test_path);
+    if (this.username === 'OpensourceJSLib') {
+      fallback_project = 'Anonymous OpenSource Project';
```
`fallback_project` variable is not being used anywhere in `config.js`

vedharish replied:
Resolved

vedharish commented:
@jzaefferer 
Updated the structure and added an example of the structure [here](https://github.com/browserstack/browserstack-runner/blob/module/README.md#usage-as-a-module)

The build is failing because it is dependent on the `js-reporter` release

jzaefferer commented:
Nice, that looks good! Now the browser reporter is even simpler, without the extra mapping.

@flore77 should be ready with a new js-reporters release today or tomorrow.

flore77 commented:
I made the new release :tada: it is on npm

akhillb commented on bin/cli.js:
```diff
@@ -386,66 +358,76 @@ var statusPoller = {
-        });
+      tunnel = new Tunnel(config.key, serverPort, config.tunnelIdentifier, config, function (err) {
+        if(err) {
+          cleanUpAndExit(null, err, {}, callback);
```
Shouldn't this be `cleanUpAndExit(null, err, [], callback)` after the new changes?

vedharish commented:
@flore77 
Hey, I bumped into one more issue.

As you can see the build failed in middle of the qunit tests. I found a console error in the browser during the test run.

[Here is a screenshot of the error](https://s3.amazonaws.com/testautomation/23da3766f0d8eb9864402d019833035f9a160172/js-screenshot-1470913478.png)

It points to `TypeError: test is undefined` in `saveTestDetails` function. 
Also, I ran the tests with the latest version of qunit and can reproduce the second error `cyclic object value`. This occurs when we try to do `JSON.stringify` on the `test` or `suite` object to `post` it to `server.js`. 

Can you please check?

flore77 commented:
@vedharish sure.

 At a first view, I tested the `tests` separately and they seem to work, see screenshot below:
![image](https://cloud.githubusercontent.com/assets/11819686/17600492/a289ac02-5ffb-11e6-9502-5b0a22c292a9.png)

flore77 commented:
I tested also with the `master` of `browserstack-runner` and the tests are passing. I think it is something in this pr that results into the error:

> TypeError: test is undefined in saveTestDetails function

flore77 commented:
Also for:

> cyclic error

![image](https://cloud.githubusercontent.com/assets/11819686/17601633/03c0a9da-6001-11e6-9551-8878d8d3641c.png)

It seems to work, also I don't know why it should show up, because we do not have circular references in our emitted objects, or at least I didn't spot anyone.

vedharish commented:
@flore77 

So, here is what I did -

- Cloned qunit repository, checked out `1.21.0` tag and ran `grunt build`
- Did `npm install js-reporters -g` and noted the installed path
- Opened up the file `test/index.html` in the qunit repo, and added 2 lines just before the `</head>` tag 

```node
<script src="/Users/harish/.nvm/versions/node/v4.3.0/lib/node_modules/js-reporters/dist/js-reporters.js"></script>
<script src="/tmp/inject.js"></script>
```

Here are the contents of `/tmp/inject.js` file -

```node
(function() {
  var runner = new JsReporters.QUnitAdapter(QUnit);

  runner.on('testEnd', function(eachTest) {
    console.log(JSON.stringify(eachTest));
  });

  runner.on('runEnd', function(globalSuite) {
    console.log(JSON.stringify(globalSuite));
  });
})();
```

Then I opened up the `test/index.html` file locally and I can reproduce both the errors. 
<img width="1680" alt="screen shot 2016-08-12 at 11 41 34 am" src="https://cloud.githubusercontent.com/assets/2497481/17614200/f23f2354-6081-11e6-8ad7-184698f32aba.png">

flore77 commented:
@vedharish thanks, I could reproduce the error now.

All problems are caused by using `JSON.stringify`, the circular problem is caused by the objects from the`Assertion`, `actual` and `expected` property. This are independent from js-reporters, they will be objects returned by the testing framework.

So I don't know if we can do much about this, from the js-reporters side. @jzaefferer what do you think about this?

This could be solved in some way, by the`replacer` function that json.stringify.

flore77 commented:
This problem was there before introducing js-reporters, in [QUnit's reporter](https://github.com/browserstack/browserstack-runner/blob/d14d04801fabb34bbd6a518fb6638b73356cdac5/lib/_patch/qunit-plugin.js#L46-L47), but it was including only the errors, so it didn't happen to be circulars objects, but now that js-reporters provides all assertions, the passed ones contains some circular objects.

jzaefferer commented:
I agree that this is something that `browserstack-runner` needs to handle. The circular structure comes from QUnit's testsuite directly and can't be decycled by js-reporters' adapter.

I found two modules for safe `JSON.stringify`:

* https://github.com/WebReflection/circular-json
* https://github.com/isaacs/json-stringify-safe

Not sure which one is better here. Here's some discussion from the WebReflection one about the isaacs one: https://github.com/WebReflection/circular-json#why-not-the-izs-one

jzaefferer commented on lib/server.js:
```diff
@@ -60,6 +80,7 @@ exports.Server = function Server(bsClient, workers) {
         });
 
         patch += '<script type="text/javascript">' + jsReportersScript + '</script>';
+        patch += '<script type="text/javascript">' + circularJSONScript + '</script>';
```
Should abstract this, something like
```
patch += externalScript('../node_modules/js-reporters/dist/js-reporters.js');
patch += externalScript('../node_modules/circular-json/build/circular-json.js');
```

jzaefferer commented on lib/server.js:
```diff
@@ -219,26 +233,35 @@ exports.Server = function Server(bsClient, workers) {
 
       try {
-        query = parseBody(body);
+        query = CircularJSON.parse(body);
```
Before, `body` was first passed through `qs.parse(body).data.escapeSpecialChars()`. Is it safe to drop that? I doubt that `CircularJSON` is handling that part.

vedharish replied:
@jzaefferer Earlier the `POST` requests were [`application/x-www-form-urlencoded`](https://github.com/browserstack/browserstack-runner/blob/master/lib/_patch/browserstack.js#L31) and the `JSON data` was passed in body [as value of `data=`](https://github.com/browserstack/browserstack-runner/blob/master/lib/_patch/browserstack.js#L26) which I changed to `application/json` and directly `POSTing` the JSON data. This way we don't have to `encodeURIComponent` and/or `use querystring and do escapeSpecialChars()`. Simple `CircularJSON.parse` will work here I think.

jzaefferer replied:
gotcha, thats much better indeed

jzaefferer commented:
Looks good to me.
